ThunkLibs/libX11/libX11_Guest.cpp

```diff
@@ -68,16 +73,44 @@ extern "C" {
         return rv;
     }
 
-    static void LockMutexFunction() {
+    static void LockMutexFunction(LockInfoPtr) {
       fprintf(stderr, "libX11: LockMutex\n");
     }
 
-    static void UnlockMutexFunction() {
+    static void UnlockMutexFunction(LockInfoPtr) {
       fprintf(stderr, "libX11: LockMutex\n");
     }
 
-  void (*_XLockMutex_fn)() = LockMutexFunction;
-  void (*_XUnlockMutex_fn)() = UnlockMutexFunction;
+  int XFree(void* ptr) {
+    // This function must be able to handle both guest heap pointers *and* host heap pointers,
+    // so it only forwards to the native host library for the latter.
+    //
+    // This is because Xlibint users allocate memory using internal macros aliasing to libc's
+    // malloc but then they free using the function XFree. For libX11, this is not a problem since
+    // the allocation happens in a thunked API function (and hence on the host heap), but if a
+    // function from an unthunked library accesses Xlibint, it will allocate on the guest heap.
+    //
+    // One notable example where this was encountered is XF86VidModeGetAllModeLines.
+
+    if (!ptr || IsHostHeapAllocation(ptr)) {
+      return fexfn_pack_XFree(ptr);
+    } else {
+      free(ptr);
+      return 1;
+    }
+  }
+
+  void XFreeEventData(Display* display, XGenericEventCookie* cookie) {
+    // Has the same heap-mismatch issue as XFree, so we have to reimplement it manually
+    if (_XIsEventCookie(display, (XEvent*)cookie) && cookie->data) {
```

Comment:
Is this the only event that ever needs to be freed? No need to free cookie itself?

Reply:
Yeah, this reflects behavior of the proper XFreeEventData implementation. I'm guessing no other kinds use the `data` field, and the cookie itself is deallocated separately.